### PR TITLE
[FLINK-21549][table-planner-blink] Support json serialization/deserialization for the push-down result of DynamicTableSource and DynamicTableSink

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Currently, Flink supports applying different kinds of ability to Source: projection push down, filter push, watermark push down, limit push down, partition push down, reading metadata. The push-down result should also be serialized to the json plan and then we can get the correct ExecNode plan from the json plan. To avoid modifying each connector, we can store the push-down result into the TableSourceTable during the Calcite optimization phase, and store the push-down result into DynamicTableSourceSpec in the ExecNode graph.  The similar thing can be done for DynamicTableSink. This pr aims to support them.*

## Brief change log

  - *Support json serialization/deserialization for the push-down result of DynamicTableSource*
  - *Support json serialization/deserialization for the push-down result of DynamicTableSink*

## Verifying this change

This change added tests and can be verified as follows:

  - *Extended DynamicTableSourceSpecSerdeTest and DynamicTableSinkSpecSerdeTest to verify the change*
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/RexNodeExtractor.scala [hunk 1]
```diff
@@ -108,18 +130,15 @@ object RexNodeExtractor extends Logging {
...
 
-    val convertedExpressions = new mutable.ArrayBuffer[Expression]
+    val convertibleRexNodes = new mutable.ArrayBuffer[RexNode]
     val unconvertedRexNodes = new mutable.ArrayBuffer[RexNode]
-    val inputNames = inputFieldNames.asScala.toArray
-    val converter = new RexNodeToExpressionConverter(
-      rexBuilder, inputNames, functionCatalog, catalogManager, timeZone)
     conjunctions.asScala.foreach(rex => {
       rex.accept(converter) match {
-        case Some(expression) => convertedExpressions += expression
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/RexNodeExtractor.scala:118]: there is a regression here: for rexnode which can not be converted to a expression, will be treated as convertible rexnode

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/abilities/source/FilterPushDownSpec.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    private final List<RexNode> predicates;
+
+    @JsonCreator
+    public FilterPushDownSpec(@JsonProperty(FIELD_NAME_PREDICATES) List<RexNode> predicates) {
+        this.predicates = new ArrayList<>(checkNotNull(predicates));
+    }
+
+    @Override
+    public void apply(DynamicTableSource tableSource, SourceAbilityContext context) {
+        apply(predicates, tableSource, context);
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/abilities/source/FilterPushDownSpec.java:62]: I think we need to check that t all predicates is accepted by the source after deserialization

**Review by wenlong88 — COMMENTED** [review]
the pr LGTM in all and makes the abilities supported in sources and sinks much more clearer in exec node, just two minor comment left.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/RexNodeExtractor.scala (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/RexNodeExtractor.scala:118]: no, convertible nodes are that can be converted via RexNodeToExpressionConverter

**Review by wenlong88 — APPROVED** [review]

**Pull request closed without merging**